voicevox_engine/tts_pipeline/tts_engine.py

```diff
@@ -374,6 +426,143 @@ def synthesize_wave(
+            if note.lyric == "":
+                if note.key is not None:
+                    raise HTTPException(
+                        status_code=400,
+                        detail="note.lyric is empty but note.key is not null",
+                    )
+                note_lengths.append(note.length)
+                note_consonants.append(-1)
+                note_vowels.append(0)  # pau
+                phonemes.append(0)  # pau
+                phoneme_keys.append(-1)
+            else:
+                if note.key is None:
+                    raise HTTPException(
+                        status_code=400,
+                        detail="note.lyric is not empty but note.key is null",
+                    )
+
+                # TODO: 1ノートに複数のモーラがある場合の処理
+                mora_phonemes = mora_kana_to_mora_phonemes.get(
+                    note.lyric  # type: ignore
+                ) or mora_kana_to_mora_phonemes.get(
+                    _hira_to_kana(note.lyric)  # type: ignore
+                )
+                if mora_phonemes is None:
+                    raise HTTPException(
+                        status_code=400,
+                        detail="note.lyric is not valid",
+                    )
+
+                consonant, vowel = mora_phonemes
+                if consonant is None:
+                    consonant_id = -1
+                else:
+                    consonant_id = Phoneme(consonant).id
+                vowel_id = Phoneme(vowel).id
+
+                note_lengths.append(note.length)
+                note_consonants.append(consonant_id)
+                note_vowels.append(vowel_id)
+                if consonant_id != -1:
+                    phonemes.append(consonant_id)
+                    phoneme_keys.append(note.key)
+                phonemes.append(vowel_id)
+                phoneme_keys.append(note.key)
+
+        # 各データをnumpy配列に変換する
+        note_lengths_array = np.array(note_lengths, dtype=np.int64)
+        note_consonants_array = np.array(note_consonants, dtype=np.int64)
+        note_vowels_array = np.array(note_vowels, dtype=np.int64)
+        phonemes_array = np.array(phonemes, dtype=np.int64)
+        phoneme_keys_array = np.array(phoneme_keys, dtype=np.int64)
+
+        # コアを用いて子音長を生成する
+        consonant_lengths = self._core.safe_predict_sing_consonant_length_forward(
+            note_consonants_array, note_vowels_array, note_lengths_array, style_id
+        )
+
+        # 予測した子音長を元に、すべての音素長を計算する
+        phoneme_lengths = calc_phoneme_lengths(consonant_lengths, note_lengths_array)
+
+        # 時間スケールを変更する（音素 → フレーム）
+        frame_phonemes = np.repeat(phonemes_array, phoneme_lengths)
+        frame_keys = np.repeat(phoneme_keys_array, phoneme_lengths)
+
+        # コアを用いて音高を生成する
+        f0s = self._core.safe_predict_sing_f0_forward(
+            frame_phonemes, frame_keys, style_id
+        )
+
+        # コアを用いて音量を生成する
+        volumes = self._core.safe_predict_sing_volume_forward(
```

Comment:
（細かいけど）複数形のsと_array取れるかも？
いやarrayの方はいるかも…。兄弟メソッドの方に合わせておくと間違いないはず！

Reply:
arrayはないとmypyに怒られたので、まあ外せないのかなと...
変数名のリファクタリングは後回しにしちゃっても特に問題はないかなと思いました。

Reply:
なるほどです。ではこちらのFIXMEコメントだけ･･･！
```suggestion
        # FIXME: 変数名のsいらない？
        volumes = self._core.safe_predict_sing_volume_forward(
```